subsys/fs/nvs/nvs.c

```diff
@@ -389,400 +755,179 @@ int nvs_init(struct nvs_fs *fs, const char *dev_name, u32_t magic)
-	 * and slot of next data
+	/* the space available should be big enough to always leave a sector
+	 * end, which is of length 0.
 	 */
-	hdr_len = _nvs_len_in_flash(fs, sizeof(struct _nvs_data_hdr));
-	slt_len = _nvs_len_in_flash(fs, sizeof(struct _nvs_data_slt));
-	extended_len = required_len + hdr_len + slt_len;
+	extended_len = required_len + _nvs_flash_size(fs, 0);
 
 	if ((fs->sector_size - (fs->write_location & (fs->sector_size - 1))) <
 		extended_len) {
 		rc = NVS_STATUS_NOSPACE;
 		goto err;
 	}
-	data_hdr.id = entry->id;
-	data_hdr.len = _nvs_len_in_flash(fs, entry->len);
-
-	rc = nvs_flash_write(fs, fs->write_location, &data_hdr, sizeof(data_hdr));
-	if (rc) {
-		goto err;
-	}
-
-	entry->data_addr = fs->write_location + hdr_len;
-	fs->write_location += required_len;
-	rc = 0;
-
-err:
-	k_mutex_unlock(&fs->nvs_lock);
-	return rc;
-}
-
-/* close the append by applying a seal to the data that includes a crc */
-int nvs_append_close(struct nvs_fs *fs, const struct nvs_entry *entry)
-{
-	int rc;
-	struct _nvs_data_slt data_slt;
-	off_t addr;
-
-	k_mutex_lock(&fs->nvs_lock, K_FOREVER);
-	rc = nvs_compute_crc(fs, entry, &data_slt.crc16);
-	if (rc) {
-		goto err;
-	}
-	addr = _nvs_slt_addr_in_flash(fs, entry);
-	rc = nvs_flash_write(fs, addr, &data_slt, sizeof(data_slt));
+	return 0;
 
 err:
-	k_mutex_unlock(&fs->nvs_lock);
 	return rc;
 }
 
-/* check the crc of an entry */
-int nvs_check_crc(struct nvs_fs *fs, struct nvs_entry *entry)
-{
-	int rc;
-	struct _nvs_data_slt data_slt;
-	off_t addr;
-	u16_t crc16;
-
-	rc = nvs_compute_crc(fs, entry, &crc16);
-	if (rc) {
-		return -EIO;
-	}
-	addr = _nvs_slt_addr_in_flash(fs, entry);
-	rc = nvs_flash_read(fs, addr, &data_slt, sizeof(data_slt));
-	if (rc || (crc16 != data_slt.crc16)) {
-		return -EIO;
-	}
-	return 0;
-}
-
-
-
 /* rotate the nvs, frees the next sector (based on fs->write_location) */
 int nvs_rotate(struct nvs_fs *fs)
 {
 	int rc;
-	off_t addr;
-	struct _nvs_data_hdr head;
-	u16_t hdr_len, slt_len, sec_hdr_len;
+	off_t offset;
+	size_t len;
+	u8_t buf[NVS_BLOCK_SIZE];
 
 	rc = 0;
 
-	k_mutex_lock(&fs->nvs_lock, K_FOREVER);
-
-	/* fill previous sector with data to jump to the next sector */
-	head.id = NVS_ID_SECTOR_END;
-	hdr_len = _nvs_len_in_flash(fs, sizeof(struct _nvs_data_hdr));
-	slt_len = _nvs_len_in_flash(fs, sizeof(struct _nvs_data_slt));
-	sec_hdr_len = _nvs_len_in_flash(fs, sizeof(struct _nvs_sector_hdr));
-	head.len = fs->sector_size
-		   - (fs->write_location & (fs->sector_size - 1))
-		   - slt_len - hdr_len + sec_hdr_len;
-	rc = nvs_flash_write(fs, fs->write_location, &head, sizeof(head));
+	len = fs->sector_size - (fs->write_location & (fs->sector_size - 1));
+	/* fill previous sector with data to jump to the next sector
+	 * the len1 calculation is only valid for smaller blocks, to avoid
+	 * errors the sector is filled with smaller items.
+	 */
+	memset(buf, 0xff, NVS_BLOCK_SIZE);
+	while (len > (NVS_BLOCK_SIZE + _nvs_flash_size(fs, 0))) {
```

Comment:
I understood that this fill procedure is introduced due the backward serach feature you had added?

Reply:
Hi andrzej, also the first version of NVS has filling at the end of a sector. The length was even set to jump over the next sector header. However only the length was written to flash. Now ff's are written. The procedure has changed because of the variable length format, making the length parameter variable in size makes it impossible to reach certain sizes (e.g. on a system with byte write size a length of 127 bytes generates a fill block of 133 bytes; a length of 128 bytes generates a length of 136 bytes, so it is impossible to generate a block of size 134 and 135). This is of course more complicated for different write sizes, so I decided to avoid this difficulty by cutting it down to smaller blocks.